d3a_api_client/redis_client_base.py

```diff
@@ -177,18 +178,27 @@ def offer_energy_rate(self, energy, rate):
         log_bid_offer_confirmation(response)
         return response
 
-    @registered_connection
-    def bid_energy(self, energy, price):
+    def bid_energy(self, energy, price, replace_existing=True):
         logging.debug(f"{self.area_id}Client tries to place a bid for {energy} kWh at {price} cents.")
-        response = self._publish_and_wait(Commands.BID, {"energy": energy, "price": price})
+
+        response = self._publish_and_wait(
+            Commands.BID,
+            {"energy": energy, "price": price, 'replace_existing': replace_existing})
+
         log_bid_offer_confirmation(response)
+
         return response
 
     @registered_connection
-    def bid_energy_rate(self, energy, rate):
+    def bid_energy_rate(self, energy, rate, replace_existing=True):
         logging.debug(f"Client tries to place a bid for {energy} kWh at {rate} cents/kWh.")
-        response = self._publish_and_wait(Commands.BID, {"energy": energy, "price": rate * energy})
+
+        response = self._publish_and_wait(
+            Commands.BID,
+            {"energy": energy, "price": rate * energy, 'replace_existing': replace_existing})
+
         log_bid_offer_confirmation(response)
+
         return response
 
     @registered_connection
```

Comment:
How about the `offer_energy` and `offer_energy_rate` ?

Reply:
I had planned to create a separate PR for `offer_energy` and `offer_energy_rate` (as I am doing on d3a). Would you rather have them together? In that case, this PR would need to wait before getting merged, until both the "bids" PR and the future "offers" PR will be merged on d3a.

Reply:
yes, please add both, as stated in the tickets title.
Also, as stated in d3a, you can test with the corresponding d3a branch without having to push it to master first.
Just adapt the dockerfile for the integrationtests. If you need more help, let me know. Thanks!